### DISCUSSION TITLE
Uploading files from NodeJs and Express

### DISCUSSION THREAD

**Discussion by nicolasggdev** [pocketbase/pocketbase#1951] [Q&A] [answered]
Hi everyone, I'm building a Rest API with Pocketbase using js-sdk, NodeJs (V.18) and Js. I have been trying to create a record from the server with an image but have not been able to make it possible. This is the code... Is it possible to do this? What could be my mistake? Thanks in advance

```
export const createNewAd = catchAsync(async (req, res, next) => {
  const {
    title,
    description,
    start_date,
    finish_date,
    total_cost,
    ad_contractor_id,
    country_id,
    province_id,
    city_id
  } = req.body;

  const { file } = req;

  let form = new FormData();
  form.append("title", title);
  form.append("description", description);
  form.append("start_date", start_date);
  form.append("finish_date", finish_date);
  form.append("total_cost", total_cost);
  form.append("ad_contractor_id", ad_contractor_id);
  form.append("status_id", codStatus.active);
  form.append("country_id", country_id);
  form.append("province_id", province_id);
  form.append("city_id", city_id);
  form.append("file", new Blob([file.buffer]), "test.png");

  await pb
    .collection("ads")
    .create(form)
    .catch((err) => console.log(err));
});
```

**Comment by ganigeorgiev** [comment]
What is `file.buffer`? I cannot find such fields in the [`req` Express docs](https://expressjs.com/en/api.html#req).

**Reply by ganigeorgiev** [reply]
From the above docs:

> In Express 4, req.files is no longer available on the req object by default. To access uploaded files on the req.files object, use multipart-handling middleware like [busboy](https://www.npmjs.com/package/busboy), [multer](https://www.npmjs.com/package/multer), [formidable](https://www.npmjs.com/package/formidable), [multiparty](https://www.npmjs.com/package/multiparty), [connect-multiparty](https://www.npmjs.com/package/connect-multiparty), or [pez](https://www.npmjs.com/package/pez).

Are you using older version of Express?

**Comment by nicolasggdev** [comment]
I'm trying to do is receive the file in the request and create the formData from the information received in the body and with the file. This is what I see in the console if I make a console.log() to the form and the file.

![Pocketbase_image](https://user-images.githubusercontent.com/91642122/222205487-005344c2-4bff-48c4-a7d8-32bff9f8bdfc.png)



I'm using Node 18.14.2 and Express 4.18.2.

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand how the statement `const { file } = req;` is resolved.
There is no such field `file` in the [`req` Express docs](https://expressjs.com/en/api.html#req).

Could you provide a more detailed code sample or minimal reproducible repo?

**Comment by nicolasggdev** [comment, marked as answer]
Yes, this is the repository: https://github.com/nicolasggdev/pocketbase_api.git

The function is located in the following path: controllers/ads.controller/createNewAd

To send information, you must use this endpoint with the HTTP POST method: http://localhost:3500/api/v1/ads

An example of postman

![Pocketbase_postman](https://user-images.githubusercontent.com/91642122/222208626-1cd5435f-a83e-4773-bf35-83373a2f9ee5.png)

**Reply by ganigeorgiev** [reply]
You have `file` prop in the `req` Express object because you are using [`multer`](https://www.npmjs.com/package/multer).

Based on your exported collections from https://github.com/nicolasggdev/pocketbase_api/blob/main/database/collections.json, you are using a wrong key in the FormData, it must be `media`, aka. change:

```js
// old
form.append("file", new Blob([file.buffer]), "test.png");

// new
form.append("media", new Blob([file.buffer]), "test.png");
```

---

Additionally, please note that you are using a single SDK instance for all node requests - https://github.com/nicolasggdev/pocketbase_api/blob/main/database/db.js. This is ok for now, since you don't have regular user authentication, but if you decide to add it later, you'll have to create a one-off client for each node request-response (eg. it could be created inside a middleware and passed as local/context prop), because otherwise requests from different clients will update the same `pb.authStore` state.

**Reply by nicolasggdev** [reply]
That was the error, the code already works hehe. I will take your recommendations into account.

Thank you so much !!
